Bluetooth: Support for HCI Vendor-Specific Events

thoh-ot opened the pull request:
This PR is the result of the discussions in https://github.com/zephyrproject-rtos/zephyr/pull/16057.

An API to register a vendor defined callback for HCI Vendor-Specific Events has been added.

When called, the vendor callback can decode and handle the event and in doing so it will signal to the stack that the event has been handled; if not the stack itself will have a possibility to decode and handle the event (currently there are no builtin handlers in the stack)

The PR also adds a test that runs the Host standalone with a stubbed test controller that only supports a very limited HCI command set.

thoh-ot committed:
tests: bluetooth: Test vendor-specific events

Add test for a proprietary HCI Vendor-Specific event.

In the test the controller is replaced with a small stub controller,
with limited support of the initialization flow from the host.

After initialization, a HCI prop event callback is registered and an
prop event is triggered and sent to the host. It is then verified that
the registered callback is called.

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3088,6 +3092,9 @@ static void hci_reset_complete(struct net_buf *buf)
 	scan_dev_found_cb = NULL;
+#if defined(CONFIG_BT_HCI_VENDOR_EVENT_CALLBACK)
+	hci_vnd_evt_cb = NULL;
+#endif /* CONFIG_BT_HCI_VENDOR_EVENT_CALLBACK */
```
This doesn't seem to be the right place to do this. The other callbacks that are set to NULL here are related to active LE roles which get implicitly disable incase HCI_Reset is sent, however a vendor event callback should be fine to have available at any point, including getting registered before initialising the reset of the stack. Since the variable is declared static I guess you don't need this explicit setting to NULL anywhere.

thoh-ot replied:
Yes, good point

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -83,6 +83,10 @@ static bt_ready_cb_t ready_cb;
 static bt_le_scan_cb_t *scan_dev_found_cb;
 
+#if defined(CONFIG_BT_HCI_VENDOR_EVENT_CALLBACK)
+static bt_hci_vnd_evt_cb_t *hci_vnd_evt_cb;
```
I don't think we use `*` anywhere for function pointers. You also don't use it e.g. in the `bt_hci_register_vnd_evt_cb` parameter type so just drop it from here as well.

thoh-ot replied:
We use right above in scan_dev_found_cb

jhedberg replied:
right (sorry for being blind) - still seems inconsistent to me. I need to take a look at the wider code base since we should probably just make all these consistent with each other.

jhedberg replied:
These actually differ in the type definition as well. Some have `typedef int (*foo)(args...);` whereas others have `typedef int foo(args...);`. Maybe that determines the necessity for the `*` in the variable declarations (my C-knowledge is a bit lacking on this aspect of function pointer syntax). Either way, I suppose we want to make this consistent throughout the code base (which is a topic for a separate PR).

thoh-ot replied:
I guess it depends if you wan't a typedef the refers to a function or to a pointer to a function. For me personally it's a bit clearer when the variable is a pointer to a function.

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3370,6 +3377,40 @@ static void le_adv_report(struct net_buf *buf)
+
+static inline void default_hci_vendor_event(struct net_buf *buf)
+{
+	/* do nothing at present time */
```
I guess it wouldn't hurt to add a BT_DBG() here about the unhandled vendor event.

thoh-ot replied:
I will add that

thoh-ot replied:
BT_DBG or BT_WARN? We use BT_WARN handle_events?

jhedberg replied:
BT_WARN sounds appropriate

jhedberg reviewed (changes requested):
The design seems mostly fine to me, just some specific style comments

thoh-ot committed:
Bluetooth: Add support for vendor-specific events

Add an event handler for HCI vendor-specific events with the
event code BT_HCI_EVT_VENDOR.

A vendor defined callback can be registered to be called when
vendor-Specific events are received in the stack. The callback can then
decode and handle the event; if not the stack will decode and handle
the event.

Signed-off-by: Thomas Ebert Hansen <thoh@oticon.com>

carlescufi commented on include/bluetooth/hci.h:
```diff
@@ -1914,6 +1915,28 @@ int bt_hci_cmd_send(u16_t opcode, struct net_buf *buf);
+  *
+  * @param buf Buffer containing event parameters.
+  *
+  * @return 0 on success or -EINVAL value on failure to handle.
```
@thoh-ot @jhedberg 
What if one wants to combine Zephyr or Mesh VS events with additional manufacturer-specific events? I would add another return value `-ENOTSUP` that then is described as "defer the handling of this vendor-specific event to Zephyr's Bluetooth stack", so that an application that wants to combine all of them can do that in a documented way, and parse only the ones that Zephyr doesn't know about, delegating the others to the stack

thoh-ot replied:
> @thoh-ot @jhedberg
> What if one wants to combine Zephyr or Mesh VS events with additional manufacturer-specific events? I would add another return value `-ENOTSUP` that then is described as "defer the handling of this vendor-specific event to Zephyr's Bluetooth stack", so that an application that wants to combine all of them can do that in a documented way, and parse only the ones that Zephyr doesn't know about, delegating the others to the stack

That is what the -EINVAL error code is supposed to mean. If the callback returns -EINVAL the stack should then look at the HCI event and see if it can decode/handle it.

I can replace -EINVAL with -ENOTSUP if that is what you want.

carlescufi replied:
Right but what about the case where it's a manufacturer event, not known to the Zephyr BLE stack, but could not be processed correctly by the hook? Maybe it was malformed, maybe it had invalid data. That's why I was proposing to have an exact check on `ENOTSUP`

carlescufi commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3370,6 +3374,42 @@ static void le_adv_report(struct net_buf *buf)
+	}
+#endif /* CONFIG_BT_HCI_VENDOR_EVENT_CALLBACK */
+
+	if (err) {
```
See  my earlier comment, I would probably do:
```
if (err == -ENOTSUP) {
   default_hci_vendor_event()
}
else {
  return err;
}

thoh-ot replied:
There is nothing to return an error code back to; you cannot indicate to the controller that the event wasn't handled.

carlescufi replied:
OK, good point. Then I'm happy if we replace EINVAL with ENOTSUP and document it properly

thoh-ot replied:
@carlescufi What about other return values then?

```
0 : hook handled the event
-ENOTSUP : defer the handling of this vendor-specific event to Zephyr's Bluetooth stack
any other value : ???? (maybe assert because the hook is not following the defined contract)
```

Or maybe we should just say that 0 mean that the hook has handled the value and any value other that 0 means "defer the handling of this vendor-specific event to Zephyr's Bluetooth stack"

jhedberg replied:
Or, just define an `enum { BT_HCI_VND_EVT_HANDLED, BT_HCI_VND_EVT_NOT_HANDLED }` and have the function return that?

carlescufi replied:
OK, in that case why don't we just use a bool as a return value?

thoh-ot replied:
I used int because all the other functions in hci.h returns int and have "@return 0 on success or negative error value on failure.", but if bool is okay as a return value then i'm in favor of that.

jhedberg replied:
> in that case why don't we just use a bool as a return value?

Readibility, I suppose. I generally try to use bool only when the name of the function is inherently boolean, i.e. it's clear from the name what true/false means. I'm not sure that's true here. That said, it's of course some overhead to add a new enum definition, so I don't have a super strong opinion on this.

jhedberg replied:
@thoh-ot in this case we're not really dealing with a "failure or succcess" case, which is what int is normally used for. Rather, both possible return values can mean that everything went as expected.

thoh-ot replied:
What about just saying:

0 : hook handled the vendor-specific event
any other value: defer the handling of this vendor-specific event to Zephyr's Bluetooth stack

carlescufi commented on include/bluetooth/hci.h:
```diff
@@ -1914,6 +1915,28 @@ int bt_hci_cmd_send(u16_t opcode, struct net_buf *buf);
+  *
+  * @return 0 on success or negative error value on failure.
+  */
+int bt_hci_register_vnd_evt_cb(bt_hci_vnd_evt_cb_t cb);
```
This is a compromise between cleanliness and optimization, but another option would be to use a weak function here. We did this here: 
https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/bluetooth/hci/h4.c#L428

This is perhaps less clean from an architectural point of view, but saves RAM and ROM.

Thoughts?

joerchan replied:
I don't like weak functions simply because of compiler compatibility, so I would prefer a callback. If we need to save RAM and ROM we should probably do that with ifdefs instead.
Also we usually use callbacks, so I'd prefer to be consistent.

thoh-ot commented:
@jhedberg @carlescufi I have pushed some of the discussed changes from yesterday.

Callback return value is now a bool.

I did not take in the weak callback symbol.

The feature is now using three Kconfig symbols (not quite sure about the naming)

- BT_HCI_VENDOR_EVENT_SUPPORT <-- (general symbol for the overall feature)
  - BT_HCI_VENDOR_EVENT_STACK <-- (enable for the Zephyr VS/Extension events)
  - BT_HCI_VENDOR_EVENT_CALLBACK <-- (enable for the user callback support)

carlescufi commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -590,3 +590,23 @@ config BT_PAGE_TIMEOUT
+	# Virtual/hidden option to make the conditions more intuitive
+	bool
+
+config BT_HCI_VENDOR_EVENT_STACK
```
I don't think we need this one at all, we already have `BT_HCI_VS_EXT`, and so that could be enabled by default since soon we will handle VS events.

thoh-ot replied:
Hmm not sure about the BT_HCI_VS_EXT.
According to the help text that is for some special HCI commands.

Also looking at BT_HCI_MESH_EXT which i guess would be a user of VS events, it doesn't even depend on BT_HCI_VS_EXT, how does that fit together?

carlescufi replied:
We can update the text. For me, `BT_HCI_VS_EXT` enables handling of the `0xff` event. After that, we need a new configuration option for the handling of additional VS events that are not part of the Zephyr set. I would propose that this is: `BT_HCI_VS_EVT_HOOK`

carlescufi replied:
A bit more clarification:

* Zephyr VS commands and events:  the ones defined in https://github.com/zephyrproject-rtos/zephyr/blob/master/doc/reference/bluetooth/hci.txt
* User VS commands: Additional commands and events defined by the user

`BT_HAS_HCI_VS`: Controller supports Zephyr VS commands and events
`BT_HCI_VS`: Host and Controller enable support for Zephyr VS commands and events
`BT_HCI_VS_EXT`: Host and Controller enable support for extended Zephyr VS commands and events
`BT_HCI_VS_EXT_DETECT`: Auto-detect if the controller supports Zephyr VS commands and events

`BT_HCI_VS_EVT_USER`: Add support for registering a user event callback

carlescufi reviewed (approved):
Looks good to me now, thank you for your patience.

joerchan reviewed (approved):
Looks good

jhedberg commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -590,3 +590,11 @@ config BT_PAGE_TIMEOUT
+	  Enable registering a callback for delegating to the user the handling of
+	  VS events that are not known to the stack
+
+
```
Two unnecessary empty lines at the end of the file

thoh-ot replied:
Okay, i'll push a fix

jhedberg commented on tests/bluetooth/hci_prop_evt/src/main.c:
```diff
@@ -0,0 +1,388 @@
+	net_buf_unref(buf);
+}
+
+
```
One unnecessary empty line

thoh-ot replied:
Okay, i'll push a fix

jhedberg commented on tests/bluetooth/hci_prop_evt/src/main.c:
```diff
@@ -0,0 +1,388 @@
+	.quirks       = BT_QUIRK_NO_RESET,
+};
+
+
```
one unnecessary empty line

thoh-ot replied:
Okay, i'll push a fix

jhedberg reviewed (approved):
Looks good. There are still white space issues (unnecessary empty lines). I'll approve but it'd be nice to fix these before merging.

thoh-ot commented:
@jhedberg Just pushed changes to your comments

jhedberg commented:
> @jhedberg Just pushed changes to your comments

@thoh-ot thanks!
